src/main/scala/uk/gov/hmrc/govukfrontend/views/viewmodels/dateinput/DateValidationSupport.scala

```diff
@@ -0,0 +1,219 @@
+      .transform(
+        { case (year, month, day) => GovUkDate(day, month, year) },
+        (govUkDate: GovUkDate) => (govUkDate.entered.year, govUkDate.entered.month, govUkDate.entered.day)
+      )
+
+  private def applyGovukDateValidation(
+    invalidDateError: String,
+    missingDateError: String,
+    missingYearError: String,
+    missingMonthError: String,
+    missingDayError: String
+  ): Mapping[(String, MonthEntered, String)] =
+    Forms.of(new Formatter[(String, MonthEntered, String)] {
+      def bind(key: String, data: Map[String, String]): Either[Seq[FormError], (String, MonthEntered, String)] =
+        (
+          data.get(s"$key.year").map(_.trim),
+          data.get(s"$key.month").map(_.trim),
+          data.get(s"$key.day").map(_.trim)
+        ) match {
+
+          case (Some(""), Some(""), Some("")) =>
+            Left(Seq(FormError(key, missingDateError)))
+
+          // TODO similar checks for any combination of two fields empty
+
+          case (Some(year), Some(month), Some(day)) =>
+            val maybeValidMonth = monthValidator.validate(month)
+
+            val validationChecks = Seq[(Boolean, String, String)](
+              (year.isEmpty, "year", missingYearError),
+              (Try(year.toInt).isFailure, "year", invalidDateError),
+              (month.isEmpty, "month", missingMonthError),
+              (maybeValidMonth.isEmpty, "month", invalidDateError),
+              (day.isEmpty, "day", missingDayError),
+              (Try(day.toInt).isFailure, "day", invalidDateError),
+              (Try(day.toInt < 1 || day.toInt > 31).getOrElse(true), "day", invalidDateError)
+            )
+
+            val fieldLevelErrors = validationChecks.collect { case (true, field, error) =>
+              FormError(s"$key.$field", error, Seq(dateDescription))
+            }
+
+            if (fieldLevelErrors.nonEmpty) {
+              val distinctFieldLevelErrors = fieldLevelErrors
+                .groupBy(_.key)
+                .flatMap { case (_, fieldErrors) => fieldErrors.headOption }
+                .toSeq
+              Left(distinctFieldLevelErrors)
+            } else {
+              Try(LocalDate.of(year.toInt, maybeValidMonth.get, day.toInt)) match {
+                case Failure(_) => Left(Seq(FormError(key, invalidDateError, Seq(dateDescription))))
+                case Success(_) => Right((year, MonthEntered(month, maybeValidMonth.get), day))
+              }
+            }
+        }
+
+      def unbind(key: String, ymd: (String, MonthEntered, String)): Map[String, String] =
+        Map(s"$key.year" -> ymd._1, s"$key.month" -> ymd._2.entered, s"$key.day" -> ymd._3)
+    })
+
+  // this version doesn't include the GOVUK-recommended error for when all 3 fields are empty
+  // and it might be difficult/impossible to add some GOVUK-recommended errors, eg. for multiple missing fields
+  // but it uses more idiomatic Play verification, which may make it more maintainable
```

Comment:
not sure if a mapping is more or less idiomatic than a formatter, just that a formatter is a lower level building block